tcp/server.go

```diff
@@ -33,113 +25,82 @@ func NewServer(h Handler) (*Server, error) {
 	}
 
 	return &Server{
-		handler:   h,
-		listeners: map[*net.Listener]struct{}{},
+		handler: h,
 	}, nil
 }
 
 // Listen listens to an address for new connections, passing them
 // off to the handler in a goroutine.
-func (s *Server) Listen(addr string) error {
-	ln, err := net.Listen("tcp", addr)
+func (s *Server) Listen(ctx context.Context, addr string) error {
+	lc := &net.ListenConfig{}
+	ln, err := lc.Listen(ctx, "tcp", addr)
 	if err != nil {
-		return err
+		return fmt.Errorf("failed to listen: %w", err)
```

Comment:
```suggestion
		return fmt.Errorf("listening: %w", err)
```

Not sure if other errors are consistent with this though?

I checked and at least some of them are. I'd suggest making this consistent at least in the files touched by this PR if you can 👍